### DISCUSSION TITLE
Calling pyanaconda directly from Python

### DISCUSSION THREAD

**Discussion by korewaChino** [rhinstaller/anaconda#3956] [Q&A] [answered]
Hello! I would like to write my own UI/frontend for Anaconda in Python. But There seems to be little to no documentation on how to use the pyanaconda API.

Where do I get started on using pyanaconda directly in Python?

**Comment by VladimirSlavik** [comment, marked as answer]
Hi! Sorry, that isn't entirely possible yet. In theory, Anaconda runs as a bunch of processes (modules) glued together by the D-Bus API. So you can have a different front-end in any language, just call the API. In practice, there are at least two reasons why that isn't true:

- There is a "main process" that still does a lot of initialization and then starts the modules. You would have to duplicate that somehow.
- The actual payload installation is also not entirely modularized yet. That means you would have to deal with something that is temporary.
- The D-bus API does not have any good documentation. We are still stuck with the above, do not know what further changes need to happen, and do not want to commit to any semblance of "this is supported" (that's mostly a RHEL thing).

As we are working towards the Web UI, we will have to solve most of these problems, so it's mostly a question of time. Then again, that time could be measured in years.

----

Alternatively:
- Where to start? Read anaconda.py in the repo root, that's the main process. There's the standard line `if __name__ == "__main__":` and everything else follows, mostly.
- Individual modules are in `pyanaconda/modules/*`. Things in `pyanaconda/core/*` are mostly helpers. GUI is in `pyanaconda/ui/gui`, likewise for TUI except s/g/t/ :) Everything else under `pyanaconda` is in various stages of being refactored (slowly), or hoping to be refactored one day. The goal is to have only `modules`, `core`, and `ui`.
- It's useful to know that there are other projects/libraries that split from Anaconda over the years: Blivet, pykickstart, dasbus.
- If you can and want, use an IDE to navigate the code. Half the team uses PyCharm, which also pseudo-understands the D-Bus API and can skip over that layer to the code in modules, which is just awesome.

**Comment by poncovka** [comment]
I would like to add that the current Anaconda API is definitely not stable nor supported at this moment.
